# Beyond the Basics: Answering Tough Questions With Data

In this video we will explore a data set of airline flight data.  This will give us a chance to practice and reinforce basic pandas concepts.  It will also allow us to get a bit deeper into grouping and group functions. These are certainly more complicated topics, but this analysis should leave you with a stronger intuition for what you can do with other data sets.

The set that we will work with is a subset of data about airline flights made available on Stat-computing.org:
http://stat-computing.org/dataexpo/2009/

In [1]:
%matplotlib inline
import sys
print(sys.version)
import numpy as np
print(np.__version__)
import pandas as pd
print(pd.__version__)
import matplotlib.pyplot as plt

3.5.0 (v3.5.0:374f501f4567, Sep 12 2015, 11:00:19) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
1.9.2
0.16.2


In this week's materials, there is a file that contains data on all flights in June 1994. Let's check the data format to see if we need to convert any types or datatimes.


In [2]:
pd.read_csv("../data/june_airplane_data.csv", header=False).head()

1994  6   5  7  1542  1540  1819  1815  US  236  ...   NA.2  NA.3  0  NA.4  \
0  1994  6   6  1  1549  1540  1831  1815  US  236  ...    NaN   NaN  0   NaN   
1  1994  6   7  2  1540  1540  1803  1815  US  236  ...    NaN   NaN  0   NaN   
2  1994  6   8  3  1541  1540  1808  1815  US  236  ...    NaN   NaN  0   NaN   
3  1994  6   9  4  1541  1540  1835  1815  US  236  ...    NaN   NaN  0   NaN   
4  1994  6  10  5  1548  1540  1821  1815  US  236  ...    NaN   NaN  0   NaN   

   0.1  NA.5 NA.6 NA.7  NA.8  NA.9  
0    0   NaN  NaN  NaN   NaN   NaN  
1    0   NaN  NaN  NaN   NaN   NaN  
2    0   NaN  NaN  NaN   NaN   NaN  
3    0   NaN  NaN  NaN   NaN   NaN  
4    0   NaN  NaN  NaN   NaN   NaN  

[5 rows x 29 columns]

We can see that our DataFrame does not include column names because these are not given in the CSV file. This is a good problem for us to have now because it lets us introduce a new method.

## Reading Data From the Web

One nice thing about pandas is that we can very easily read tables on web pages. The `read_html` method looks through an html file for `<table>` tags and provides a list of all these tables as DataFrames. For example, below is the table of names, taken from the Stat-computing.org web page.

In [3]:
pd.read_html("http://stat-computing.org/dataexpo/2009/the-data.html")[1]

0                  1                                                  2
0  NaN               Name                                        Description
1    1               Year                                          1987-2008
2    2              Month                                               1-12
3    3         DayofMonth                                               1-31
4    4          DayOfWeek                            1 (Monday) - 7 (Sunday)
5    5            DepTime                actual departure time (local, hhmm)
6    6         CRSDepTime             scheduled departure time (local, hhmm)
7    7            ArrTime                  actual arrival time (local, hhmm)
8    8         CRSArrTime               scheduled arrival time (local, hhmm)
9    9      UniqueCarrier                                unique carrier code
10  10          FlightNum                                      flight number
11  11            TailNum                                  plane tail number
12  12  ActualElapsedTime                                         in minutes
13  13     CRSElapsedTime                                         in minutes
14  14            AirTime                                         in minutes
15  15           ArrDelay                          arrival delay, in minutes
16  16           DepDelay                        departure delay, in minutes
17  17             Origin                           origin IATA airport code
18  18               Dest                      destination IATA airport code
19  19           Distance                                           in miles
20  20             TaxiIn                           taxi in time, in minutes
21  21            TaxiOut                           taxi out time in minutes
22  22          Cancelled                          was the flight cancelled?
23  23   CancellationCode  reason for cancellation (A = carrier, B = weat...
24  24           Diverted                                    1 = yes, 0 = no
25  25       CarrierDelay                                         in minutes
26  26       WeatherDelay                                         in minutes
27  27           NASDelay                                         in minutes
28  28      SecurityDelay                                         in minutes
29  29  LateAircraftDelay                                         in minutes

We really only need the first column, so let's grab that in the same statement.

In [4]:
pd.read_html("http://stat-computing.org/dataexpo/2009/the-data.html")[1][1]

0                  Name
1                  Year
2                 Month
3            DayofMonth
4             DayOfWeek
5               DepTime
6            CRSDepTime
7               ArrTime
8            CRSArrTime
9         UniqueCarrier
10            FlightNum
11              TailNum
12    ActualElapsedTime
13       CRSElapsedTime
14              AirTime
15             ArrDelay
16             DepDelay
17               Origin
18                 Dest
19             Distance
20               TaxiIn
21              TaxiOut
22            Cancelled
23     CancellationCode
24             Diverted
25         CarrierDelay
26         WeatherDelay
27             NASDelay
28        SecurityDelay
29    LateAircraftDelay
Name: 1, dtype: object

We can set the names of our DataFrame columns to this series when we read in our CSV.

In [5]:
df = pd.read_csv('../data/june_airplane_data.csv', names=pd.read_html("http://stat-computing.org/dataexpo/2009/the-data.html")[1][1])

In [6]:
df.head()

Name  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  1994     6      5           7       1542     1540        1819     1815   
1  1994     6      6           1       1549     1540        1831     1815   
2  1994     6      7           2       1540     1540        1803     1815   
3  1994     6      8           3       1541     1540        1808     1815   
4  1994     6      9           4       1541     1540        1835     1815   

  CRSArrTime  UniqueCarrier        ...          TaxiIn  TaxiOut  Cancelled  \
0         US            236        ...             NaN        0        NaN   
1         US            236        ...             NaN        0        NaN   
2         US            236        ...             NaN        0        NaN   
3         US            236        ...             NaN        0        NaN   
4         US            236        ...             NaN        0        NaN   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0                 0       NaN           NaN          NaN      NaN   
1                 0       NaN           NaN          NaN      NaN   
2                 0       NaN           NaN          NaN      NaN   
3                 0       NaN           NaN          NaN      NaN   
4                 0       NaN           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            NaN                NaN  

[5 rows x 30 columns]

Unfortunately we can see that our columns are off by one. This is because we do not have a name column in the data set. The fix is straightforward; all we need to do is shift over the names list by one. We can do this by simply selecting the values that we want with a list slice. 

In [7]:
df = pd.read_csv('../data/june_airplane_data.csv', names=pd.read_html("http://stat-computing.org/dataexpo/2009/the-data.html")[1][1][1:])

In [8]:
df.head()

Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
0  1994      6           5          7     1542        1540     1819   
1  1994      6           6          1     1549        1540     1831   
2  1994      6           7          2     1540        1540     1803   
3  1994      6           8          3     1541        1540     1808   
4  1994      6           9          4     1541        1540     1835   

   CRSArrTime UniqueCarrier  FlightNum        ...          TaxiIn  TaxiOut  \
0        1815            US        236        ...             NaN      NaN   
1        1815            US        236        ...             NaN      NaN   
2        1815            US        236        ...             NaN      NaN   
3        1815            US        236        ...             NaN      NaN   
4        1815            US        236        ...             NaN      NaN   

   Cancelled  CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0          0               NaN         0           NaN          NaN      NaN   
1          0               NaN         0           NaN          NaN      NaN   
2          0               NaN         0           NaN          NaN      NaN   
3          0               NaN         0           NaN          NaN      NaN   
4          0               NaN         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay  
0            NaN                NaN  
1            NaN                NaN  
2            NaN                NaN  
3            NaN                NaN  
4            NaN                NaN  

[5 rows x 29 columns]

Now that everything lines up correctly, let's get into the data.

In [9]:
df.describe()

Year   Month     DayofMonth      DayOfWeek        DepTime  \
count  426490  426490  426490.000000  426490.000000  423805.000000   
mean     1994       6      15.520458       3.923185    1343.071375   
std         0       0       8.670839       1.930781     474.274010   
min      1994       6       1.000000       1.000000       1.000000   
25%      1994       6       8.000000       2.000000     930.000000   
50%      1994       6      16.000000       4.000000    1330.000000   
75%      1994       6      23.000000       6.000000    1731.000000   
max      1994       6      30.000000       7.000000    2400.000000   

          CRSDepTime        ArrTime     CRSArrTime      FlightNum  TailNum  \
count  426490.000000  422641.000000  426490.000000  426490.000000        0   
mean     1337.525309    1489.413914    1493.821825     901.690736      NaN   
std       465.902464     489.868444     474.554100     593.512783      NaN   
min         5.000000       1.000000       1.000000       1.000000      NaN   
25%       930.000000    1114.000000    1117.000000     407.000000      NaN   
50%      1328.000000    1517.000000    1520.000000     792.000000      NaN   
75%      1722.000000    1913.000000    1910.000000    1429.000000      NaN   
max      2359.000000    2400.000000    2400.000000    3219.000000      NaN   

             ...          TaxiIn  TaxiOut      Cancelled  CancellationCode  \
count        ...               0        0  426490.000000                 0   
mean         ...             NaN      NaN       0.006296               NaN   
std          ...             NaN      NaN       0.079095               NaN   
min          ...             NaN      NaN       0.000000               NaN   
25%          ...             NaN      NaN       0.000000               NaN   
50%          ...             NaN      NaN       0.000000               NaN   
75%          ...             NaN      NaN       0.000000               NaN   
max          ...             NaN      NaN       1.000000               NaN   

            Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
count  426490.000000             0             0         0              0   
mean        0.002729           NaN           NaN       NaN            NaN   
std         0.052171           NaN           NaN       NaN            NaN   
min         0.000000           NaN           NaN       NaN            NaN   
25%         0.000000           NaN           NaN       NaN            NaN   
50%         0.000000           NaN           NaN       NaN            NaN   
75%         0.000000           NaN           NaN       NaN            NaN   
max         1.000000           NaN           NaN       NaN            NaN   

       LateAircraftDelay  
count                  0  
mean                 NaN  
std                  NaN  
min                  NaN  
25%                  NaN  
50%                  NaN  
75%                  NaN  
max                  NaN  

[8 rows x 26 columns]

Now we have a new issue: we have a problem with our `datetime`. They are in multiple fields, and for us to more easily query this data, they should be in one.

There are two ways to resolve this. We can do it either when we read in the data or when we have already got the data read in. I will show you the former and leave the latter as an exercise for you to try yourself. The first way involves adding some parameters to `read_csv`. What we will do is specify a new `Date` field that is a combination of several other fields. This is entered into our `read_csv` function as a dictionary.

In [10]:
df = pd.read_csv('../data/june_airplane_data.csv', parse_dates={'Date':['Year','Month','DayofMonth']}, names=pd.read_html("http://stat-computing.org/dataexpo/2009/the-data.html")[1][1][1:])

That will automatically parse the combination of those columns for us and print it out correctly. We can see that it is a bit slow; however, it is actually a very convenient way of manipulating the data in a repeatable way.

In [11]:
df.head()

Date  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0 1994-06-05          7     1542        1540     1819        1815   
1 1994-06-06          1     1549        1540     1831        1815   
2 1994-06-07          2     1540        1540     1803        1815   
3 1994-06-08          3     1541        1540     1808        1815   
4 1994-06-09          4     1541        1540     1835        1815   

  UniqueCarrier  FlightNum  TailNum  ActualElapsedTime        ...          \
0            US        236      NaN                157        ...           
1            US        236      NaN                162        ...           
2            US        236      NaN                143        ...           
3            US        236      NaN                147        ...           
4            US        236      NaN                174        ...           

   TaxiIn  TaxiOut  Cancelled  CancellationCode Diverted CarrierDelay  \
0     NaN      NaN          0               NaN        0          NaN   
1     NaN      NaN          0               NaN        0          NaN   
2     NaN      NaN          0               NaN        0          NaN   
3     NaN      NaN          0               NaN        0          NaN   
4     NaN      NaN          0               NaN        0          NaN   

   WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
0           NaN       NaN            NaN                NaN  
1           NaN       NaN            NaN                NaN  
2           NaN       NaN            NaN                NaN  
3           NaN       NaN            NaN                NaN  
4           NaN       NaN            NaN                NaN  

[5 rows x 27 columns]

## Binning and Categorical Data

When we perform analysis on data, it can be useful to convert numerical data into categorical data. More specifically, we create a list of `bins` to drop a range of data into. This is different than grouping because we are converting from a continuous measure to a categorical. However, keep in mind that this procedure causes a loss of granularity and also introduces potential bias from the person creating the bins.

Let's see a real-world example of binning in practice. We will create some time bins because right now the departure and arrival times are in military format. It would be nice to break up the day into 6-hour chunks from midnight to 6 a.m., 6 a.m. to noon, noon to 1800, and finally 1800 to midnight. This is converting a continuous variable to a categorical one.

One thing to keep in mind is that when we are performing this binning, we are looking at the *actual* departure/arrival times, not the scheduled ones.

I will start by creating a range of values effectively creating the edges of my bins along with the relevant labels.

In [12]:
ranges = [0,600,1200,1800,2400]
labels = ['Early Morning','Morning','Early Afternoon','Evening']

Now we use the cut function to cut them up into groups.

In [13]:
?pd.cut()

We will call our new variable DepTime2.

In [14]:
df['DepTime2'] = pd.cut(df.DepTime, ranges, labels=labels).astype('category')

We can do the same with the arrival time.

In [15]:
df['ArrTime2'] = pd.cut(df.ArrTime, ranges, labels=labels).astype('category')

In [16]:
df.ArrTime2.head()

0    Evening
1    Evening
2    Evening
3    Evening
4    Evening
Name: ArrTime2, dtype: category
Categories (4, object): [Early Morning < Morning < Early Afternoon < Evening]

You can see that we set the dtype for our new columns to be `category`.  This is different than plain `object` data, which is how Python reads strings by default. The category dtype provides an ordering so that some things are less than others. This is needed for some types of analysis.

In [17]:
df[['DepTime2','ArrTime2']].describe()

DepTime2         ArrTime2
count    423805           422641
unique        4                4
top     Morning  Early Afternoon
freq     163514           156806

When we describe our new variables, we can see the top count as well as its frequency. Notice how this `describe` function is a bit different from what we are used to. 

We have played around with the data a bit, but we should start thinking about specific questions that we want to answer. Obviously airplane delays are common, but what is the average delay time? More specifically, is the average delay time longer in the early morning? In the afternoon? Are departure or arrival delays longer?

Let's group the data to answer these questions.

We can begin with a `groupby` in which we group our new categorical variables together, then select the relevant columns and calculate their average. Since we are concerned with delay times, we will pull out those columns for examination.

In [18]:
df.groupby(['DepTime2', 'ArrTime2'])[['DepDelay','ArrDelay']].mean()

DepDelay   ArrDelay
DepTime2        ArrTime2                             
Early Morning   Early Morning    43.804627  42.069923
                Morning           2.578566   2.109582
                Early Afternoon        NaN        NaN
                Evening          64.000000  61.333333
Morning         Early Morning          NaN        NaN
                Morning           2.367489   0.664017
                Early Afternoon   3.769618   4.102513
                Evening           5.411565  12.923469
Early Afternoon Early Morning     8.890995  14.450237
                Morning           4.896104   2.480519
                Early Afternoon   5.166529   4.571237
                Evening           7.632955  11.082200
Evening         Early Morning    36.956347  42.037799
                Morning           5.464702   7.629198
                Early Afternoon  29.142857  22.500000
                Evening          13.668750  12.756813

This table was easy to generate and contains a lot of useful information.  We can now answer questions like the following: Given that the departure time is in the early morning, what are the average arrival delays if it is to land in the morning or evening?

When I first saw this query, I asked myself why I was getting those `NaN` values. How could it be possible that we have no flights that land in the early afternoon after leaving in the early morning? This is due to our having only a subset of the data--not all of it. We can run a sanity check (always a good idea) by running the following command. *PLEASE CONFIRM THAT "NaN" SHOULD BE IN COURIER. (IT WAS LIKE THAT ALREADY.)*

In [19]:
sum((df.DepTime < 600) & (df.ArrTime > 1200) & (df.ArrTime < 1800))

0

This is a simple SQL-like query that yields the result that we are expecting. Now let's continue answering more questions.  We have a bunch of average delay times, but how much variation is there in the delays?  Let's add in the standard deviations using the `agg` method.

In [20]:
df.groupby(['DepTime2', 'ArrTime2']).agg({'ArrDelay':[np.mean, np.std], 'DepDelay':[np.mean, np.std]})

DepDelay              ArrDelay           
                                      mean        std       mean        std
DepTime2        ArrTime2                                                   
Early Morning   Early Morning    43.804627  91.836738  42.069923  91.241073
                Morning           2.578566  16.595306   2.109582  18.436171
                Evening          64.000000   8.888194  61.333333   5.507571
Morning         Morning           2.367489  10.216042   0.664017  13.453101
                Early Afternoon   3.769618  12.482494   4.102513  17.172703
                Evening           5.411565  15.704599  12.923469  24.192638
Early Afternoon Early Morning     8.890995  27.347542  14.450237  42.653210
                Morning           4.896104   8.718680   2.480519  11.084140
                Early Afternoon   5.166529  14.861790   4.571237  18.710938
                Evening           7.632955  19.324282  11.082200  28.073863
Evening         Early Morning    36.956347  65.635363  42.037799  74.136225
                Morning           5.464702  12.298888   7.629198  17.129908
                Early Afternoon  29.142857  32.229602  22.500000  30.933552
                Evening          13.668750  39.551075  12.756813  34.126398

The standard deviation measures the spread around the mean of the data. We can see that flights that leave in the early afternoon and arrive in the early morning seem to have highly variable arrival delays: the standard deviation is over three times the mean. This could be useful information for the next time you fly.

## Ranking

Let's try and answer another question: Which flights are worst for delays? To answer this question, we will learn about ranking.  Specifically, how can we rank the flights by those that have, on average, the worst delay time? Which flight in each departure time group has the worst delay time?

In [21]:
df.head()

Date  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0 1994-06-05          7     1542        1540     1819        1815   
1 1994-06-06          1     1549        1540     1831        1815   
2 1994-06-07          2     1540        1540     1803        1815   
3 1994-06-08          3     1541        1540     1808        1815   
4 1994-06-09          4     1541        1540     1835        1815   

  UniqueCarrier  FlightNum  TailNum  ActualElapsedTime    ...     Cancelled  \
0            US        236      NaN                157    ...             0   
1            US        236      NaN                162    ...             0   
2            US        236      NaN                143    ...             0   
3            US        236      NaN                147    ...             0   
4            US        236      NaN                174    ...             0   

   CancellationCode  Diverted  CarrierDelay WeatherDelay NASDelay  \
0               NaN         0           NaN          NaN      NaN   
1               NaN         0           NaN          NaN      NaN   
2               NaN         0           NaN          NaN      NaN   
3               NaN         0           NaN          NaN      NaN   
4               NaN         0           NaN          NaN      NaN   

   SecurityDelay  LateAircraftDelay         DepTime2  ArrTime2  
0            NaN                NaN  Early Afternoon   Evening  
1            NaN                NaN  Early Afternoon   Evening  
2            NaN                NaN  Early Afternoon   Evening  
3            NaN                NaN  Early Afternoon   Evening  
4            NaN                NaN  Early Afternoon   Evening  

[5 rows x 29 columns]

To rank our flights, we are going to create a function that takes in a DataFrame and returns a DataFrame that ranks by a certain column. Before we can do that, we are going to create a column of the total delay time. We are also going to need to create a new column that bins the scheduled departure times according to our categorical ordering above so that we can create those groupings.

In [22]:
df['TotalDelay'] = df.ArrDelay + df.DepDelay
df['ScheduledDepTime'] = pd.cut(df.CRSDepTime, ranges, labels=labels)

Now let's create our average delay function.

In [23]:
def average_delay(dataframe):
    dataframe['AvgFlightDelay'] = dataframe.TotalDelay.mean()
    return dataframe

As is usually the case, there is more than one way for us to proceed.  Below, I will demonstrate two methods: a simple one and a more compact one.  Your choice really depends on the information you want to get from the data.

In [24]:
avg_delayed = df.groupby('FlightNum').apply(average_delay)
avg_delayed.head()
# also 
# df.groupby('FlightNum').agg({"TotalDelay":np.mean}).sort('TotalDelay')

Date  DayOfWeek  DepTime  CRSDepTime  ArrTime  CRSArrTime  \
0 1994-06-05          7     1542        1540     1819        1815   
1 1994-06-06          1     1549        1540     1831        1815   
2 1994-06-07          2     1540        1540     1803        1815   
3 1994-06-08          3     1541        1540     1808        1815   
4 1994-06-09          4     1541        1540     1835        1815   

  UniqueCarrier  FlightNum  TailNum  ActualElapsedTime       ...        \
0            US        236      NaN                157       ...         
1            US        236      NaN                162       ...         
2            US        236      NaN                143       ...         
3            US        236      NaN                147       ...         
4            US        236      NaN                174       ...         

   CarrierDelay  WeatherDelay  NASDelay  SecurityDelay LateAircraftDelay  \
0           NaN           NaN       NaN            NaN               NaN   
1           NaN           NaN       NaN            NaN               NaN   
2           NaN           NaN       NaN            NaN               NaN   
3           NaN           NaN       NaN            NaN               NaN   
4           NaN           NaN       NaN            NaN               NaN   

          DepTime2  ArrTime2  TotalDelay  ScheduledDepTime  AvgFlightDelay  
0  Early Afternoon   Evening           6   Early Afternoon       13.608696  
1  Early Afternoon   Evening          25   Early Afternoon       13.608696  
2  Early Afternoon   Evening         -12   Early Afternoon       13.608696  
3  Early Afternoon   Evening          -6   Early Afternoon       13.608696  
4  Early Afternoon   Evening          21   Early Afternoon       13.608696  

[5 rows x 32 columns]

`avg_delayed` gives us the average delay for each flight number, which we can see in the far right column. Now let's create our more abstract ranking function. This should take in a DataFrame and a column name and rank it by that column.

In [25]:
def ranking(dataframe, column):
    dataframe.sort(column, ascending=False, inplace=True)
    dataframe[column + "Rank"] = np.arange(len(dataframe)) + 1
    return dataframe

Now let's rank our flights by average delay.  We will use the apply function again in order to complete this, and we will be grouping by our scheduled departure times.

In [26]:
avg_delayed_ranked = avg_delayed.groupby('ScheduledDepTime').apply(lambda x: ranking(x, 'AvgFlightDelay'))

Now let's take a look at the top ranking.

In [27]:
avg_delayed_ranked[avg_delayed_ranked.AvgFlightDelayRank == 1]

Date  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
ScheduledDepTime                                                              
Early Morning    269530 1994-06-02          4       51         110      109   
Morning          267020 1994-06-28          2     1158        1151     1253   
Early Afternoon  60742  1994-06-12          7     1605        1600     1703   
Evening          266946 1994-06-09          4     1953        2005     2022   

                         CRSArrTime UniqueCarrier  FlightNum  TailNum  \
ScheduledDepTime                                                        
Early Morning    269530         131            DL       1490      NaN   
Morning          267020        1250            DL       1428      NaN   
Early Afternoon  60742         1659            US       2849      NaN   
Evening          266946        2039            DL       1428      NaN   

                         ActualElapsedTime         ...          WeatherDelay  \
ScheduledDepTime                                   ...                         
Early Morning    269530                 18         ...                   NaN   
Morning          267020                 55         ...                   NaN   
Early Afternoon  60742                  58         ...                   NaN   
Evening          266946                 29         ...                   NaN   

                         NASDelay  SecurityDelay  LateAircraftDelay  \
ScheduledDepTime                                                      
Early Morning    269530       NaN            NaN                NaN   
Morning          267020       NaN            NaN                NaN   
Early Afternoon  60742        NaN            NaN                NaN   
Evening          266946       NaN            NaN                NaN   

                                DepTime2         ArrTime2  TotalDelay  \
ScheduledDepTime                                                        
Early Morning    269530    Early Morning    Early Morning         -41   
Morning          267020          Morning  Early Afternoon          10   
Early Afternoon  60742   Early Afternoon  Early Afternoon           9   
Evening          266946          Evening          Evening        1411   

                         ScheduledDepTime  AvgFlightDelay  AvgFlightDelayRank  
ScheduledDepTime                                                               
Early Morning    269530     Early Morning       36.875000                   1  
Morning          267020           Morning      173.780142                   1  
Early Afternoon  60742    Early Afternoon      183.333333                   1  
Evening          266946           Evening      173.780142                   1  

[4 rows x 33 columns]

There are some issues with this analysis because flight numbers repeat. We can see 1428 is both the worst in the morning and the evening, even though we specified depature time. This could be because the data is represented strangely or that the flight departure time changed. This is a great place to continue exploring to understand the data better.

## Transformations

We have seen how the `apply` method allows us to modify entire DataFrames. There are many situations, however, in which you only want to apply a function at the column level. The way to do that is with the `transform` method. 

To demonstrate this method in action, let's take our numerical variables and get the z-score of each observation. The z-score or standard score is the number of standard deviations from the mean of a specific value. It is an easy way of normalizing a variable so that it appears on a standard scale for comparison. You can read more here:

https://en.wikipedia.org/wiki/Standard_score

Note that our variables probably are not normally distributed, but a z-score still contains useful information about each observation. Let's write out our formula as a lambda function and then apply it using transform.

In [28]:
zscore = lambda x: (x - x.mean()) / x.std()

In [29]:
df.groupby('DayOfWeek').transform(zscore)

DepTime  CRSDepTime   ArrTime  CRSArrTime  FlightNum  TailNum  \
0       0.371092    0.385981  0.628003    0.632618  -1.110867      NaN   
1       0.426897    0.430839  0.690721    0.673286  -1.124572      NaN   
2       0.423158    0.432551  0.645238    0.674042  -1.128920      NaN   
3       0.417945    0.431557  0.651190    0.674650  -1.126593      NaN   
4       0.410037    0.433709  0.700429    0.675495  -1.127322      NaN   
5       0.427105    0.433554  0.672448    0.675952  -1.127048      NaN   
6       0.481837    0.499775  0.723677    0.737666  -1.105561      NaN   
7       0.576642    0.385981  0.833285    0.632618  -1.110867      NaN   
8       0.410150    0.430839  0.668534    0.673286  -1.124572      NaN   
9       0.425277    0.432551  0.684458    0.674042  -1.128920      NaN   
10      0.438900    0.431557  0.699869    0.674650  -1.126593      NaN   
11      0.428751    0.433709  0.686479    0.675495  -1.127322      NaN   
12      0.605069    0.433554  0.869392    0.675952  -1.127048      NaN   
13      0.486162    0.499775  0.730031    0.737666  -1.105561      NaN   
14      0.366809    0.385981       NaN    0.632618  -1.110867      NaN   
15      0.408057    0.430839  0.692738    0.673286  -1.124572      NaN   
16      0.427396    0.432551  0.703036    0.674042  -1.128920      NaN   
17      0.415850    0.431557  0.663360    0.674650  -1.126593      NaN   
18      0.412116    0.433709  0.656586    0.675495  -1.127322      NaN   
19      0.422918    0.433554  0.674479    0.675952  -1.127048      NaN   
20      0.481837    0.499775  0.700380    0.737666  -1.105561      NaN   
21      0.377515    0.385981  0.628003    0.632618  -1.110867      NaN   
22      0.565060    0.430839  0.702822    0.673286  -1.124572      NaN   
23      0.429516    0.432551  0.684458    0.674042  -1.128920      NaN   
24      1.061262    0.431557  1.352977    0.674650  -1.126593      NaN   
25      1.071288    0.433709  1.523483    0.675495  -1.127322      NaN   
26      1.633332    1.672168  1.452363    1.493111  -1.124886      NaN   
27      1.632728    1.673452  1.469676    1.493639  -1.125617      NaN   
28      1.628885    1.671176  1.464282    1.492411  -1.125342      NaN   
29      1.642937    1.649011  1.465719    1.467068  -1.109226      NaN   
...          ...         ...       ...         ...        ...      ...   
426460  1.264525    1.276456  1.097413    1.104867   1.406848      NaN   
426461  1.229323    1.278017  1.045195    1.105545   1.403551      NaN   
426462  1.237364    1.276417  1.064305    1.105116   1.404414      NaN   
426463  1.210424    1.246148  1.026126    1.071239   1.324973      NaN   
426464  1.251852    1.276542  1.066339    1.103928   1.406225      NaN   
426465       NaN    1.276456       NaN    1.104867   1.406848      NaN   
426466  1.245958    1.278017  1.077081    1.105545   1.403551      NaN   
426467  1.411141    1.276417  1.242975    1.105116   1.404414      NaN   
426468  1.218988    1.246148  1.034420    1.071239   1.324973      NaN   
426469  1.208529    1.224580  1.041209    0.969188   1.403551      NaN   
426470  1.208053    1.223071  1.048062    0.969040   1.404414      NaN   
426471  1.150472    1.191707  0.893418    0.932164   1.324973      NaN   
426472  1.182606    1.220889  1.035619    0.966773   1.403493      NaN   
426473  1.279404    1.223125  1.099367    0.967623   1.406225      NaN   
426474  1.185656    1.224580  1.029252    0.969188   1.403551      NaN   
426475  0.210491    0.206964  0.099496    0.076544   1.406848      NaN   
426476  0.208334    0.209273  0.102569    0.077621   1.403551      NaN   
426477  0.156976    0.157329  0.032893    0.022827   1.324973      NaN   
426478  0.217559    0.206636  0.127992    0.075830   1.403493      NaN   
426479  0.213335    0.208198  0.098219    0.076394   1.406225      NaN   
426480  0.332030    0.206964  0.136005    0.076544   1.406848      NaN   
426481  0.210414    0.209273  0.096591    0.077621   1.403551      NaN   
426482  0.196

One thing you will notice is some of the transformed variables do not make sense; normalizing flight number does not make any sense because technically this variable is categorical, not numerical. We can specify specific columns for transformation in the manner we have seen before.

Now we have had a little practice applying pandas to real data, and you should be developing some intuition for the types of operations that will help you in different scenarios. The skills you are learning are going to be useful, but they take practice to master.  Make sure you take some time to analyze your own data sets using the tools covered here.